### PR TITLE
Asynchronous loadflow calculation

### PR DESCRIPTION
**Pull request by anistouri**

### CODE REVIEW

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 1]
```diff
@@ -642,15 +645,48 @@ public ResponseEntity<Void> moveOrCopyModifications(@PathVariable("studyUuid") U
...
+        String result = loadflowService.getLoadFlowResult(nodeUuid);
+        return result != null ? ResponseEntity.ok().body(result) :
+                ResponseEntity.noContent().build();
+    }
+
+    @GetMapping(value = "/studies/{studyUuid}/nodes/{nodeUuid}/loadflow/status")
+    @Operation(summary = "Get the loadflow status on study")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The loadflow status"),
+            @ApiResponse(responseCode = "204", description = "No loadflow has been done yet"),
+            @ApiResponse(responseCode = "404", description = "The loadflow status has not been found")})
```
Comment (flomillot) [inline src/main/java/org/gridsuite/study/server/StudyController.java:672]:
```suggestion
            @ApiResponse(responseCode = "404", description = "The loadflow has not been found")})
```
To be coherent with above

**Review by flomillot — APPROVED** [review]

File: src/main/java/org/gridsuite/study/server/service/LoadFlowService.java [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        HttpHeaders headers = new HttpHeaders();
+        headers.set(HEADER_USER_ID, userId);
+        headers.setContentType(MediaType.APPLICATION_JSON);
+
+        HttpEntity<LoadFlowParametersInfos> httpEntity = new HttpEntity<>(loadflowParameters, headers);
+
+        return restTemplate.exchange(loadFlowServerBaseUri + path, HttpMethod.POST, httpEntity, UUID.class).getBody();
+    }
+
+    public void deleteLoadFlowResult(UUID uuid) {
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/study/server/service/LoadFlowService.java:105]: This method should be called : 
- when the study is deleted 
- when the node is deleted or invalidated
- when the LF parameters are modified
You can look where is called deleteSensitivityAnalysisResult()

**Review by etiennehomer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/study/server/repository/LoadFlowResultEntity.java [hunk 3]
```diff
@@ -1,53 +0,0 @@
...
-    @GeneratedValue(strategy = GenerationType.AUTO)
-    @Column(name = "id")
-    private UUID id;
-
-    @Column(name = "ok")
-    private boolean ok;
-
-    @Column(name = "metrics")
-    @ElementCollection
-    @CollectionTable(foreignKey = @ForeignKey(name = "loadFlowResultEntity_metrics_fk"))
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/study/server/repository/LoadFlowResultEntity.java:42]: metrics are not stored in the new DB loadlow. It was not used ? We are sure that metrics are useless now ?

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java [hunk 4]
```diff
@@ -176,28 +171,20 @@ public UUID getReportUuid(UUID nodeUuid) {
         return getNode(nodeUuid).getReportUuid();
     }
 
-    public LoadFlowStatus getLoadFlowStatus(UUID nodeUuid) {
-        return getLoadFlowStatus(getNode(nodeUuid));
-    }
-
     public void updateLoadFlowResultAndStatus(UUID nodeUuid, LoadFlowResult loadFlowResult, LoadFlowStatus loadFlowStatus) {
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java:174]: To be removed

**Review by etiennehomer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/study/server/repository/LoadFlowResultEntity.java (hunk 3 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/study/server/repository/LoadFlowResultEntity.java:42]: This field is never used and it is always empty in the response from Powsybl

File: src/main/java/org/gridsuite/study/server/service/LoadFlowService.java (hunk 2 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/study/server/service/LoadFlowService.java:106]: done

File: src/main/java/org/gridsuite/study/server/repository/LoadFlowResultEntity.java (hunk 3 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/study/server/repository/LoadFlowResultEntity.java:42]: Ok to remove it

**Comment by etiennehomer** [comment]
That would be nice to test setLoadFlowParameters() method in StudyService. To increase coverage and for robustness

File: src/main/java/org/gridsuite/study/server/service/StudyService.java [hunk 5]
```diff
@@ -760,10 +760,21 @@ public String getAllMapData(UUID studyUuid, UUID nodeUuid, List<String> substati
                 substationsIds, "all");
     }
 
-    public void runLoadFlow(UUID studyUuid, UUID nodeUuid) {
+    public UUID runLoadFlow(UUID studyUuid, UUID nodeUuid, String userId) {
+        /*StudyEntity studyEntity = studyRepository.findById(studyUuid).orElseThrow(() -> new StudyException(STUDY_NOT_FOUND));
```
[the remark this answers is not in the document]
Comment (anistouri) [reply src/main/java/org/gridsuite/study/server/service/StudyService.java]: done

File: src/test/java/org/gridsuite/study/server/LoadFlowTest.java [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+
+    private static final String NETWORK_UUID_STRING = "38400000-8cf0-11bd-b23e-10b96e4ef00d";
+
+    private static final String LOADFLOW_RESULT_UUID = "1b6cc22c-3f33-11ed-b878-0242ac120002";
+
+    private static final String LOADFLOW_ERROR_RESULT_UUID = "25222222-9994-4e55-8ec7-07ea965d24eb";
+
+    private static final String LOADFLOW_OTHER_NODE_RESULT_UUID = "11131111-8594-4e55-8ef7-07ea965d24eb";
+
+    private static final String LOADFLOW_RESULT_JSON = "{\"resultUuid\":\"0c8de370-3e6c-4d72-b292-d355a97e0d5d\",\"writeTimeStamp\":\"2023-06-13T14:07:30.856705+02:00\",\"componentResults\":[{\"componentResultUuid\":\"4244c84f-855b-4935-a02e-fa34832cf55b\",\"connectedComponentNum\":1,\"synchronousComponentNum\":2,\"status\":\"CONVERGED\",\"iterationCount\":3,\"slackBusId\":\"slackBusId1\",\"slackBusActivePowerMismatch\":4.0,\"distributedActivePower\":5.0},{\"componentResultUuid\":\"83ccb778-1738-40e7-96f8-a341fa054d24\",\"connectedComponentNum\":1,\"synchronousComponentNum\":2,\"status\":\"CONVERGED\",\"iterationCount\":3,\"slackBusId\":\"slackBusId1\",\"slackBusActivePowerMismatch\":4.0,\"distributedActivePower\":5.0}]}\n";
```
Comment (flomillot) [inline src/test/java/org/gridsuite/study/server/LoadFlowTest.java:95]: We should not do that, it's not readable and hard to maintain, can you store it in a JSON file instead ?

File: src/test/java/org/gridsuite/study/server/LoadFlowTest.java [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+        assertTrue(TestUtils.getRequestsDone(1, server).stream().anyMatch(r -> r.matches("/v1/results/" + LOADFLOW_RESULT_UUID)));
+
+        // get loadflow status
+        mockMvc.perform(get("/v1/studies/{studyUuid}/nodes/{nodeUuid}/loadflow/status", studyNameUserIdUuid, modificationNode3Uuid)).andExpectAll(
+                status().isOk(),
+                content().string(LOADFLOW_STATUS_JSON));
+
+        assertTrue(TestUtils.getRequestsDone(1, server).stream().anyMatch(r -> r.matches("/v1/results/" + LOADFLOW_RESULT_UUID + "/status")));
+
+        // stop short circuit analysis
```
Comment (flomillot) [inline src/test/java/org/gridsuite/study/server/LoadFlowTest.java:263]: Short circuit -> loadflow ?

File: src/test/java/org/gridsuite/study/server/NetworkModificationTest.java [hunk 8]
```diff
@@ -2050,43 +2051,58 @@ public void testDuplicateModification() throws Exception {
...
+                        studyUuid, UUID.randomUUID())
                         .contentType(MediaType.APPLICATION_JSON)
                         .content(modificationUuidListBody)
                         .header(USER_ID_HEADER, "userId"))
                 .andExpect(status().isNotFound());
 
+    }
+
+    @Test
+    public void testCutAndPasteModification() throws Exception {
```
Comment (flomillot) [inline src/test/java/org/gridsuite/study/server/NetworkModificationTest.java:2085]: Why we have these tests in this PR ?

**Review by flomillot — CHANGES_REQUESTED** [review]

File: src/test/java/org/gridsuite/study/server/LoadFlowTest.java (hunk 7 above)
Comment (anistouri) [reply src/test/java/org/gridsuite/study/server/LoadFlowTest.java]: Done

File: src/test/java/org/gridsuite/study/server/NetworkModificationTest.java (hunk 8 above)
Comment (anistouri) [reply src/test/java/org/gridsuite/study/server/NetworkModificationTest.java:2085]: It was when I merged with the main branch and had conflicts

File: src/main/java/org/gridsuite/study/server/service/LoadFlowService.java [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+ * @author Kevin Le Saulnier <kevin.lesaulnier at rte-france.com>
+ */
+@Service
+public class LoadFlowService {
+
+    static final String RESULT_UUID = "resultUuid";
+    static final String RESULTS_UUIDS = "resultsUuids";
+    private String loadFlowServerBaseUri;
+
+    @Autowired
```
[the remark this answers is not in the document]
Comment (anistouri) [reply src/main/java/org/gridsuite/study/server/service/LoadFlowService.java]: Done

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**